File: battleship.c
// user takes turns against existing board.
// for simplicity, this only models one move.
// also for now the board being guessed about is fixed.
// 10 x 10 board, ships with the Hasboro naming scheme

#include <stdio.h>
#include <stdlib.h>

int** make_board() {
    int size = 10;
    // allocate board space
    int* start = calloc(size * size, sizeof(int));
    int** board = malloc(size * sizeof(int*));
    for (int i = 0; i < size; i++)
        board[i] = start + i * size;

    // set board to all 0s
    for (int i = 0; i < size; i++)
        for (int j = 0; j < size; j++)
            board[i][j] = 0;

    // add ships
    board[1][1] = 1; 
    board[2][1] = 1; // patrol boat
    board[9][1] = 1; 
    board[9][2] = 1;
    board[9][2] = 1; // submarine
    board[1][4] = 1; 
    board[1][5] = 1;
    board[1][6] = 1; // destroyer
    board[6][6] = 1; 
    board[6][7] = 1;
    board[6][8] = 1;
    board[6][9] = 1; // battleship
    board[3][2] = 1; 
    board[4][2] = 1;
    board[5][2] = 1;
    board[6][2] = 1;
    board[7][2] = 1; // carrier
    return board;
}

void free_board(int** board) {
    free(*board);
    free(board);
}

int check_hit(int x, int y, int** board) {
    // treat off-board shots as misses
    if (x < 0 || x > 9 || y < 0 || y > 9) return 0;
    return board[x][y];
}

int main() {
    int** board = make_board();
    int x_guess, y_guess;

    int hit = check_hit(x_guess, y_guess, board);
    // printf("%d", hit);
    free_board(board);
    return 0;
}

File: battleship.py
#!/usr/bin/env python3

# user takes turns against existing board.
# for simplicity, this only models one move.
# also for now the board being guessed about is fixed.
# 10 x 10 board, ships with the Hasboro naming scheme

def make_board():
    size = 10
    # initialize board with all 0s
    board = [[0 for _ in range(size)] for _ in range(size)]
    
    # add ships
    board[1][1] = 1
    board[2][1] = 1  # patrol boat
    
    board[9][1] = 1
    board[9][2] = 1  # submarine (note: there was a duplicate line in C version)
    
    board[1][4] = 1
    board[1][5] = 1
    board[1][6] = 1  # destroyer
    
    board[6][6] = 1
    board[6][7] = 1
    board[6][8] = 1
    board[6][9] = 1  # battleship
    
    board[3][2] = 1
    board[4][2] = 1
    board[5][2] = 1
    board[6][2] = 1
    board[7][2] = 1  # carrier
    
    return board

def check_hit(x, y, board):
    # treat off-board shots as misses
    if x < 0 or x > 9 or y < 0 or y > 9:
        return 0
    return board[x][y]

def main():
    board = make_board()
    
    # Get user input for coordinates
    try:
        x_guess = int(input("Enter X coordinate (0-9): "))
        y_guess = int(input("Enter Y coordinate (0-9): "))
        
        hit = check_hit(x_guess, y_guess, board)
        print("Hit!" if hit else "Miss!")
    except ValueError:
        print("Please enter valid numbers")

if __name__ == "__main__":
    main() 